README.md
# Description

This repository provides the iOS sdk version of the Zender Player.

Current version is `2.1.0` , target :ios 9.0 (PhenixSdk version 2019.2.1)

# Find your configuration in the Zender Admin
- Logging to the admin <https://admin.zender.tv>
- Select a specific stream
- Get the information (the I icon)
- Read the targetId and channelId

 ![Zender TargetId and ChannelId](docs/images/targetId-channelId.png?raw=true "Find your Zender TargetId and ChannelId")

# Sample project
A sample project is provided in this repo [Demo Project](ZenderDemo)

- You'll need your targetId and channelId configured in the ViewController.m
- the project uses Cocoapods , so requires a `pod install` to install all the dependencies

# Usage

```objective-c
//
//  ViewController.m
//  ZenderDemo
//
//  Created by Patrick Debois on 30/06/2019.
//  Copyright © 2019 Zender. All rights reserved.
//

#import "ViewController.h"

@import Zender;

@interface ViewController ()<ZenderPlayerDelegate>

@property (atomic, strong) ZenderPlayer *player; // ZenderPlayer is UIViewController
@property (atomic, strong) ZenderAuthentication *authentication;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Initialize the target and channel
    NSString *channelId ;
    NSString *targetId;
    
    targetId = @"<your-target-id>"; // fill-in your own targetId
    channelId = @"<your-channel-id>"; // fill-in your own channelId

    // Create a Zender Player
    self.player= [ZenderPlayer new];
    
    // Create a player configuration
    ZenderPlayerConfig* settingsConfig = [ZenderPlayerConfig configWithTargetId:targetId channelId:channelId];
    self.player.config = settingsConfig;

    [self setupAuthentication];

    self.player.authentication = _authentication;

	// Set this class as a ZenderPlayerDelegate
    self.player.delegate = self;
    
    self.player.view.frame = self.view.frame;
    self.player.view.hidden = false;
    
	[self.player start];
    
    [self.view addSubview:self.player.view];
    
    // Do any additional setup after loading the view, typically from a nib.
}

#pragma mark Zender Authentication
- (void)setupAuthentication {
    // Configure authentication, uncomment depending on your configuration

    /* No authentication 
    _authentication=nil;
    */

	/* Device Authentication
   _authentication = [ZenderAuthentication authenticationWith:@{
	    @"token": [[[UIDevice currentDevice] identifierForVendor] UUIDString],
	    @"name": @"demo",
	    @"avatar": @"https://cdn.iconscout.com/public/images/icon/free/png-512/avatar-user-hacker-3830b32ad9e0802c-512x512.png"
    } provider:@"device"];
	*/ 

	/* Facebook authentication , requires the facebook framework to be integrated
	_authentication = [ZenderAuthentication authenticationWith:@{
		@"token": [FBSDKAccessToken currentAccessToken].tokenString,
	} provider:@"facebook"];
    */

	/* Google authentication , requires the google framework to be integrated
    _authentication = [ZenderAuthentication authenticationWith:@{
        @"token": <googleToken>
    } provider:@"google"];
    */

    /*
    _authentication = [ZenderAuthentication authenticationWith:@{
    @"token": signed_json_string
    } provider:@"signedProvider"];
    */
}

#pragma mark Zender Player Delegate

// Listen for player ready event
- (void)zenderPlayer:(ZenderPlayer *)zenderPlayer onZenderPlayerReady:(NSDictionary *)payload {
}

// Listen for player close event
- (void)zenderPlayer:(ZenderPlayer *)zenderPlayer onZenderPlayerClose:(NSDictionary *)payload {
    // stop the player
    [self.player stop];
    self.player.delegate = nil;
    self.player = nil;
}

// Listen for player logout event
- (void)zenderPlayer:(ZenderPlayer *)zenderPlayer onZenderPlayerAuthenticationLogout:(NSDictionary *)payload {
    // stop the player
    [self.player stop];
    self.player.delegate = nil;
    self.player = nil;
}

// Listen for player authentication required
- (void)zenderPlayer:(ZenderPlayer *)zenderPlayer onZenderPlayerAuthenticationRequired:(NSDictionary *)payload {
    // do something to login the user
}

// Listen for Quiz Share codes
- (void)zenderPlayer:(ZenderPlayer *)zenderPlayer onZenderQuizShareCode:(NSDictionary *)payload {
    // This will open up the native share window
    [zenderPlayer sharePayload:payload controller:self];
}


@end
```

For a more detailed explantion on authentication see <https://github.com/zendertv/zender-docs/blob/master/AUTH.md>
# Installation
## Using cocoapods
The sdk is provided via private Cocoapod repository.
Add the following to the top of your Podfile

```
source 'https://github.com/zendertv/Specs.git'`
source 'https://github.com/CocoaPods/Specs.git'
```


Add this pod :
```
 # Pods for ZenderDemo
 pod 'Zender'
```

Note: The pod downloads a large zip file containing the frameworks , this might take some time

## Manual installation

A zip file containing the sdk frameworks can be downloaded from <https://repo.zender.tv/ios/zender-ios-sdk-v2.1.0.zip>

Note: the framework files are large, if you want to check them in you will require github Large File Support

### Add to Embedded Frameworks to the project
The second step is to add the frameworks as embedded frameworks:
- Select your Application Target (on the left)
- Select the General tab
- Select Add the framework via the `Embedded Binaries` (+ button)
- Select Other to add the framework
- Browse to your manually downloaded sdk file
- Add frameworks `Zender`, `ZenderPhenix` and `PhenixSdk`
- When a screen pops up to 'Choose options for adding these files' `(Select Copy if needed)`

![Add to Embedded Frameworks](docs/images/ios/framework-embed.png?raw=true "Add to Embedded Frameworks")

### Basic frameworks
Now add the necessary frameworks under "General/Linked Frameworks and Libraries": 
- UIKit.framework
- JavascriptCore.framework
- AVFoundation.framework
- AVKit.framework

# Additional configuration

### Disable bitcode
Zender depends on frameworks that are currrently not BITCODE enabled.  Therefore you need to disable it:
- Select your Application Target (on the left)
- Select the `Build Settings` tab
- Find `Enable Bitcode` and select NO

![Disable bitcode](docs/images/ios/disable-bitcode.png?raw=true "Disable bitcode")

### Background audio
To be able to play audio in background, add this to the background mode:

- Select your Application Target (on the left)
- Select the `Capabilities` tab
- Select `Audio, Airplay, and Picture in Picture`

![Background Audio](docs/images/ios/background-audio.png?raw=true "Background Audio")

### Orientation Portrait
The Zender player autorotates, if you don't want this behaviour you need to fix the app rotation or the controller

- Select your Application Target (on the left)
- Select the `General` tab
- Select/Deselect the required `Device Orientation` options

![Fix portrait](docs/images/ios/portrait-only.png?raw=true "Fix Portrait Modus")

### UI Fullscreen
On iPad screens can be viewed in split screen , it's best to disable that
![Enable Full screen only](docs/images/ios/full-screen.png?raw=true "Enable Full Screen")

## Strip frameworks
The Zender frameworks provides `armv7, arm64, x86_64` architecture builds.
To publish an app to the appstore, you need to strip the simulator part.

[A sample strip script](scripts/strip-framework.sh)


# Push notifications
The Zender Admin console provides a way to send push notifications to users to notifiy them when new streams are available.
This requires push notification certificate setup to match the bundleId of your app and allowing us to send the push notifications.

## Enable Push notifications in your project/provisioning profile
![Enable Push notifications](docs/images/ios/enable-push-notifications.png?raw=true "Enable Push notifications")


## Sample code
Just add the following part to your `AppDelegate.m`

```objective-c

@import Zender;

- (void)application:(UIApplication *)application didRegisterForRemoteNotificationsWithDeviceToken:(NSData *)deviceToken {
    [[Zender sharedInstance] application:application didRegisterForRemoteNotificationsWithDeviceToken:deviceToken];
}
```

# Deeplinking
Zender supports using Universal Deeplinks. The default zender domain is links.zender.tv . To make this work, the AppId needs to be configured in the backend.


## Enable Universal Links
Add `links.zender.tv` under your domains.

![Enable Universal Links](docs/images/ios/enable-universallinks.png?raw=true "Enable Universal links")

## Sample code
Just add the following code so Zender handles the deeplinks

```objective-c
@import Zender;

- (BOOL)application:(UIApplication *)application continueUserActivity:(NSUserActivity *)userActivity restorationHandler:(void (^)(NSArray<id<UIUserActivityRestoring>> * _Nullable))restorationHandler {
    
    BOOL handled = [[Zender sharedInstance] handleUniversalLink:userActivity.webpageURL completion:^(NSError *error) {
        
    }];
    
    return handled;
    
}

```

# Advanced
More advanced topics are described at [Advanced Documentation](ADVANCED.md)

# Version 1 Documentation
Since version 2.0.0 a lot has changed , but you can still read the [Version 1 Documentation](README-v1.md)

# Versioning
We adhere to the Semantic Versions scheme to indiciate releases. Versions starting with 0.x should be considered beta versions and not yet ready for production use.

# Privacy
Before using people their name, avatar and other information, they should give a consent to use their data.
It is up the implementor of the SDK to get the agreement of the user before exposing him in the Zender Player.

The data is securely stored on the Zender platform but there needs to be an agreement

# License
- All Zender Code is copyrighted and owned by Small Town Heroes BVBA. Extensions have their own license based on the license of the supplier.
- This code does not grant you ownership or a license to use it. This needs to be arranged by a specific license agreement.
- All code provided should be treated as confidential until official cleared.

# Changelog
See [Changelog](CHANGELOG.md)



CHANGELOG.md
# Changelog
2.1.0:
- [player] autodetect video players
- [push notification] provide application handler for easiers integration
- [deeplinking] an application handler is now available for detecting zender universal links 

1.0.5:
- [phenix] fix memory leak , better warning message

1.0.4:
- [player] feedback if video connection falls back to audio only or worse
- [player] no vibration when you have given a wrong answer

1.0.3:
- [player] fix of UserDevice weak attributes 
- [player] fix allowFileAccessFromFileURLs is a boolean not a string
- [player] catch [NSJSONSerialization dataWithJSONObject exceptions

1.0.2:
- [player] correctly set username on logging
1.0.1:
- [player] log the app version in all cases

1.0.0:
- [player] more logging
- [player] improved background/foreground spinner
- [player] workaround for phenix sdk memory leak, fix in sdk pending
- [player] correctly expose some missing events

0.9.13:
- [player] introduce remote logging
- [player] completely stop the stream in background
- [player] add new action shoutboxEnable, shoutboxDisable
- [player] added dependency on Systems.framework for network quality monitoring
- [player] new events onZenderEventShoutboxShoutSent,  onZenderEventQuizStart, onZenderEventQuizStop, ZenderEventQuizExtralifeUse, ZenderEventQuizExtralifeIgnore, ZenderEventQuizAnswerSubmit, ZenderEventQuizAnswerTimeout, ZenderEventQuizEliminated, ZenderEventQuizEliminatedContinue, ZenderEventAnswerCorrect , ZenderEventAnswerIncorrect

0.9.11:
- [player] fix statusbar animation
- [player] add vibrate on quiz question

0.9.10:
- [player] fix loader & avplayer aspectfill on iPhoneX
- [player] new event ZenderOnLoaderTimeout to detect potential issues

0.9.9:
- [player] fix for iPhoneX keyboard offset & correct screensize with saferareas
- [player] fix for correct display when statusbar is visible or not
- [player] remote logging over the bridge
- [player] disable display timer to avoid going to sleep during quiz
- [player] remove the Done keyboard button

0.9.8:
- [notifications] documentation of push payload
- [player] improvement to handle keyboard hide/show
- [player] improved phenix buffering visualisation
- [player] imrpoved medialaan documentation

0.9.7:
- [player] new callbacks onZenderQuizShareCode, onZenderMediaPlay, onZenderPlayerLobbyEnter, onZenderPlayerLobbyLeave, onZenderOpenUrl, onZenderUiStreamsOverview
- [logger] introduced logger object with different loglevels , added logging to API & Player
- [player] privacy, rules etc.. in settings now open in separate Safari view
- [player] Share code is now available in the onZenderQuizShareCode
- [debug] If debug is enabled, a long tap on the player will trigger a onZenderPlayerAdsShow event
- [player] fix inconsistency for registerDevice / registerDeviceToken on Player
- [player] deprecated `[ self.player registerDevice:<devicetoken>,  self.player unregisterDevice:<devicetoken> ]`
- [player] added a zender_loop.mp4 to the Zender/Core resources bundle
- [player] JS logging is now redirected to the ZenderLogger (tag=js) 
- [youtube] first provider for Youtube video support
- [player] improved buffering status
- [player] Added `videoStop` `videoStart` player methods. Useful for older devices that can't play with videoPause/videoResume
- [player] Added `onZenderLoaderHideoaderShow` `onZenderLoaderHide` events to allow for custom loaders 

0.9.6:
Changes:
- API Calls now follow Error First rule

0.9.5:
Changes:
- [Player] is now a UIViewController and longer a UIView
- [Player] stop command cleans all internal objects
- [Player][Phenix] video mode is now cover
- [Player] the statusbar is now removed
Bugfixes:
- [Player][Phenix][bugfix] phenix player is now full portrait
New Features:
- [Player] pauseVideo , resumeVideo added
- [Player] mute, unmute added
- [API] subscribe, unsubscribe a device
- [Player][Config] allows you to pass a device(token) to the PlayerConfig without subscribe/unsubscribe
- [Player] show loader until media is ready (experimental)

0.9.4:
- [API] Introduced ZenderApiClient to do API calls without starting the player
- [API] Push notifications can now be registered, unregistered without starting the player
- [API] getStreams allows you to retrieve the metadata of the current streams (ZenderStream)
- [API] login call now returns a ZenderSession
- [Player] Added onPlayerClose event to react on the close button in the player
- [Phenix] A beta version that works inside a simulator is now available
- [Auth] 
 mechanism for Medialaan is now available

0.9.1 : 
- First public beta version



scripts/strip-framework.sh
# Stripping framework only for archive
if [ "$ACTION" = "install" ]; then

    SCRIPT_FILE_NAME="strip-framework.sh"

    # Set working directory to product’s embedded frameworks
    FRAMEWORK_NAME="Zender"
    if [ -d "${BUILT_PRODUCTS_DIR}/${FRAMEWORKS_FOLDER_PATH}/${FRAMEWORK_NAME}.framework" ]; then
	    cd "${BUILT_PRODUCTS_DIR}/${FRAMEWORKS_FOLDER_PATH}/${FRAMEWORK_NAME}.framework"

	    # Get architectures for current file
	    ARCHS="$(lipo -info "${FRAMEWORK_NAME}" | rev | cut -d ':' -f1 | rev)"
	    for ARCH in $ARCHS; do
		if ! [[ "${VALID_ARCHS}" == *"$ARCH"* ]]; then
		# Strip non-valid architectures in-place
		lipo -remove "$ARCH" -output "$FRAMEWORK_NAME" "$FRAMEWORK_NAME" || exit 1
		fi
	    done
    	echo "Framework ${FRAMEWORK_NAME} was successfully stripped with unsupported architectures"
   fi 

    FRAMEWORK_NAME="ZenderPhenix"
    if [ -d "${BUILT_PRODUCTS_DIR}/${FRAMEWORKS_FOLDER_PATH}/${FRAMEWORK_NAME}.framework" ]; then
	    cd "${BUILT_PRODUCTS_DIR}/${FRAMEWORKS_FOLDER_PATH}/${FRAMEWORK_NAME}.framework"

	    # Get architectures for current file
	    ARCHS="$(lipo -info "${FRAMEWORK_NAME}" | rev | cut -d ':' -f1 | rev)"
	    for ARCH in $ARCHS; do
		if ! [[ "${VALID_ARCHS}" == *"$ARCH"* ]]; then
		# Strip non-valid architectures in-place
		lipo -remove "$ARCH" -output "$FRAMEWORK_NAME" "$FRAMEWORK_NAME" || exit 1
		fi
	    done
    	echo "Framework ${FRAMEWORK_NAME} was successfully stripped with unsupported architectures"
	fi

    FRAMEWORK_NAME="PhenixSdk"
    if [ -d "${BUILT_PRODUCTS_DIR}/${FRAMEWORKS_FOLDER_PATH}/${FRAMEWORK_NAME}.framework" ]; then
	    cd "${BUILT_PRODUCTS_DIR}/${FRAMEWORKS_FOLDER_PATH}/${FRAMEWORK_NAME}.framework"

	    # Get architectures for current file
	    ARCHS="$(lipo -info "${FRAMEWORK_NAME}" | rev | cut -d ':' -f1 | rev)"
	    for ARCH in $ARCHS; do
		if ! [[ "${VALID_ARCHS}" == *"$ARCH"* ]]; then
		# Strip non-valid architectures in-place
		lipo -remove "$ARCH" -output "$FRAMEWORK_NAME" "$FRAMEWORK_NAME" || exit 1
		fi
	    done
    	echo "Framework ${FRAMEWORK_NAME} was successfully stripped with unsupported architectures"
 	fi 

    FRAMEWORK_NAME="ZenderYoutube"
    if [ -d "${BUILT_PRODUCTS_DIR}/${FRAMEWORKS_FOLDER_PATH}/${FRAMEWORK_NAME}.framework" ]; then
	    cd "${BUILT_PRODUCTS_DIR}/${FRAMEWORKS_FOLDER_PATH}/${FRAMEWORK_NAME}.framework"

	    # Get architectures for current file
	    ARCHS="$(lipo -info "${FRAMEWORK_NAME}" | rev | cut -d ':' -f1 | rev)"
	    for ARCH in $ARCHS; do
		if ! [[ "${VALID_ARCHS}" == *"$ARCH"* ]]; then
		# Strip non-valid architectures in-place
		lipo -remove "$ARCH" -output "$FRAMEWORK_NAME" "$FRAMEWORK_NAME" || exit 1
		fi
	    done
    	echo "Framework ${FRAMEWORK_NAME} was successfully stripped with unsupported architectures"
	fi

    # Removing script from framework folder
	if [ -f ${SCRIPT_FILE_NAME} ]; then
		rm -rf "${SCRIPT_FILE_NAME}"
	fi

fi



ZenderDemo/ZenderDemo/AppDelegate.m
//
//  AppDelegate.m
//  ZenderDemo
//
//  Created by Patrick Debois on 30/06/2019.
//  Copyright © 2019 Zender. All rights reserved.
//

#import "AppDelegate.h"
@import Zender;

@import UserNotifications;


@interface AppDelegate ()

@end

@implementation AppDelegate


- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
    // Override point for customization after application launch.
    
    [self askForPermission];

    return YES;
}


- (void)applicationWillResignActive:(UIApplication *)application {
    // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
    // Use this method to pause ongoing tasks, disable timers, and invalidate graphics rendering callbacks. Games should use this method to pause the game.
}


- (void)applicationDidEnterBackground:(UIApplication *)application {
    // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
    // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
}


- (void)applicationWillEnterForeground:(UIApplication *)application {
    // Called as part of the transition from the background to the active state; here you can undo many of the changes made on entering the background.
}


- (void)applicationDidBecomeActive:(UIApplication *)application {
    // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
}


- (void)applicationWillTerminate:(UIApplication *)application {
    // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
}

- (void)application:(UIApplication *)application didRegisterForRemoteNotificationsWithDeviceToken:(NSData *)deviceToken {
    [[Zender sharedInstance] application:application didRegisterForRemoteNotificationsWithDeviceToken:deviceToken];
}

- (BOOL)application:(UIApplication *)application continueUserActivity:(NSUserActivity *)userActivity restorationHandler:(void (^)(NSArray<id<UIUserActivityRestoring>> * _Nullable))restorationHandler {
    
    BOOL handled = [[Zender sharedInstance] handleUniversalLink:userActivity.webpageURL completion:^(NSError *error) {
        
    }];
    
    return handled;
    
}

// Convenience method for asking for push notification permissions
- (void)askForPermission {
    
    // Override point for customization after application launch.
    // Check Notification Settings on launch
    if (@available(iOS 10.0, *)) {
        [[UNUserNotificationCenter currentNotificationCenter] getNotificationSettingsWithCompletionHandler:^(UNNotificationSettings * _Nonnull settings) {
            switch (settings.authorizationStatus) {
                    // This means we have not yet asked for notification permissions
                    case UNAuthorizationStatusNotDetermined:
                {
                    [[UNUserNotificationCenter currentNotificationCenter] requestAuthorizationWithOptions:(UNAuthorizationOptionSound | UNAuthorizationOptionAlert | UNAuthorizationOptionBadge) completionHandler:^(BOOL granted, NSError * _Nullable error) {
                        // You might want to remove this, or handle errors differently in production
                        NSAssert(error == nil, @"There should be no error");
                        if (granted) {
                            dispatch_async(dispatch_get_main_queue(), ^(void){
                                [[UIApplication sharedApplication] registerForRemoteNotifications];
                            });
                            
                        }
                    }];
                }
                    break;
                    // We are already authorized, so no need to ask
                    case UNAuthorizationStatusAuthorized:
                {
                    // Just try and register for remote notifications
                    dispatch_async(dispatch_get_main_queue(), ^(void){
                        [[UIApplication sharedApplication] registerForRemoteNotifications];
                    });
                }
                    break;
                    // We are denied User Notifications
                    case UNAuthorizationStatusDenied:
                {
                    
                    // Possibly display something to the user
                    UIAlertController *useNotificationsController = [UIAlertController alertControllerWithTitle:@"Turn on notifications" message:@"This app needs notifications turned on for the best user experience" preferredStyle:UIAlertControllerStyleAlert];
                    UIAlertAction *goToSettingsAction = [UIAlertAction actionWithTitle:@"Go to settings" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                        [[UIApplication sharedApplication] openURL:[NSURL URLWithString:UIApplicationOpenSettingsURLString]];
                    }];
                    UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleDefault handler:nil];
                    [useNotificationsController addAction:goToSettingsAction];
                    [useNotificationsController addAction:cancelAction];
                    dispatch_async(dispatch_get_main_queue(), ^(void){
                        [self.window.rootViewController presentViewController:useNotificationsController animated:false completion:nil];
                    });
                    NSLog(@"We cannot use notifications because the user has denied permissions");
                    
                }
                    break;
                    // The application is provisionally authorized to post noninterruptive user notifications.
                    case UNAuthorizationStatusProvisional: {
                        // iOS 12.0 we don't worry about it now
                        break;
                    }
            }
            
        }];
    } else {
        // Fallback on earlier versions
        // Register the supported interaction types.
        UIUserNotificationType types = UIUserNotificationTypeBadge |
        UIUserNotificationTypeSound | UIUserNotificationTypeAlert;
        UIUserNotificationSettings *mySettings =
        [UIUserNotificationSettings settingsForTypes:types categories:nil];
        [[UIApplication sharedApplication] registerUserNotificationSettings:mySettings];
        
    }
}

@end


ZenderDemo/ZenderDemoTests/ZenderDemoTests.m
//
//  ZenderDemoTests.m
//  ZenderDemoTests
//
//  Created by Patrick Debois on 30/06/2019.
//  Copyright © 2019 Zender. All rights reserved.
//

#import <XCTest/XCTest.h>

@interface ZenderDemoTests : XCTestCase

@end

@implementation ZenderDemoTests

- (void)setUp {
    // Put setup code here. This method is called before the invocation of each test method in the class.
}

- (void)tearDown {
    // Put teardown code here. This method is called after the invocation of each test method in the class.
}

- (void)testExample {
    // This is an example of a functional test case.
    // Use XCTAssert and related functions to verify your tests produce the correct results.
}

- (void)testPerformanceExample {
    // This is an example of a performance test case.
    [self measureBlock:^{
        // Put the code you want to measure the time of here.
    }];
}

@end


ZenderDemo/ZenderDemoUITests/ZenderDemoUITests.m
//
//  ZenderDemoUITests.m
//  ZenderDemoUITests
//
//  Created by Patrick Debois on 30/06/2019.
//  Copyright © 2019 Zender. All rights reserved.
//

#import <XCTest/XCTest.h>

@interface ZenderDemoUITests : XCTestCase

@end

@implementation ZenderDemoUITests

- (void)setUp {
    // Put setup code here. This method is called before the invocation of each test method in the class.

    // In UI tests it is usually best to stop immediately when a failure occurs.
    self.continueAfterFailure = NO;

    // UI tests must launch the application that they test. Doing this in setup will make sure it happens for each test method.
    [[[XCUIApplication alloc] init] launch];

    // In UI tests it’s important to set the initial state - such as interface orientation - required for your tests before they run. The setUp method is a good place to do this.
}

- (void)tearDown {
    // Put teardown code here. This method is called after the invocation of each test method in the class.
}

- (void)testExample {
    // Use recording to get started writing UI tests.
    // Use XCTAssert and related functions to verify your tests produce the correct results.
}

@end


ZenderDemo/ZenderDemo/ViewController.m

//  ViewController.m
//  ZenderDemo
//
//  Created by Patrick Debois on 30/06/2019.
//  Copyright © 2019 Zender. All rights reserved.
//

#import "ViewController.h"

@import Zender;

@interface ViewController ()<ZenderPlayerDelegate>

@property (atomic, strong) ZenderPlayer *player; // ZenderPlayer is UIViewController
@property (atomic, strong) ZenderAuthentication *authentication;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Initialize the target and channel
    NSString *channelId ;
    NSString *targetId;
    
    targetId = @"<your-target-id>"; // fill-in your own targetId
    channelId = @"<your-channel-id>"; // fill-in your own channelId
        
    // Create a Zender Player
    self.player= [ZenderPlayer new];
    
    // Create a player configuration
    ZenderPlayerConfig* settingsConfig = [ZenderPlayerConfig configWithTargetId:targetId channelId:channelId];
    self.player.config = settingsConfig;
    [settingsConfig enableDebug:TRUE];
    
    [self setupAuthentication];
    
    self.player.authentication = _authentication;
    
    // Set this class as a ZenderPlayerDelegate
    self.player.delegate = self;
    
    // Set a background color, the player is transparent by default
    self.player.view.backgroundColor=[UIColor blackColor];
    
    self.player.view.frame = self.view.frame;
    self.player.view.hidden = false;
    
    [self.player start];
    
    [self.view addSubview:self.player.view];
    
    // Do any additional setup after loading the view, typically from a nib.
}

#pragma mark Zender Authentication
- (void)setupAuthentication {
    // Configure authentication, uncomment depending on your configuration
    
    /* No authentication , handle later with onZenderPlayerAuthentication Required
     _authentication=nil;
     */
    
    /* Device Authentication , creates a user based on a unique deviceId
     _authentication = [ZenderAuthentication authenticationWith:@{
     @"token": [[[UIDevice currentDevice] identifierForVendor] UUIDString],
     @"name": @"demo",
    // @"avatar": @"https://cdn.iconscout.com/public/images/icon/free/png-512/avatar-user-hacker-3830b32ad9e0802c-512x512.png"
     } provider:@"device"];
     */
    
    /* Facebook authentication , requires the facebook framework to be integrated
     _authentication = [ZenderAuthentication authenticationWith:@{
     @"token": [FBSDKAccessToken currentAccessToken].tokenString,
     } provider:@"facebook"];
     */
    
    /* Google authentication , requires the google framework to be integrated
     _authentication = [ZenderAuthentication authenticationWith:@{
     @"token": <googleToken>
     } provider:@"google"];
     */
    
    /*
     _authentication = [ZenderAuthentication authenticationWith:@{
     @"token": signed_json_string
     } provider:@"signedProvider"];
     */
}

#pragma mark Zender Player Delegate

// Listen for player ready event
- (void)zenderPlayer:(ZenderPlayer *)zenderPlayer onZenderPlayerReady:(NSDictionary *)payload {
}

// Listen for player close event
- (void)zenderPlayer:(ZenderPlayer *)zenderPlayer onZenderPlayerClose:(NSDictionary *)payload {
    // stop the player
    [self.player stop];
    self.player.delegate = nil;
    self.player = nil;
}

// Listen for player logout event
- (void)zenderPlayer:(ZenderPlayer *)zenderPlayer onZenderPlayerAuthenticationLogout:(NSDictionary *)payload {
    // stop the player
    [self.player stop];
    self.player.delegate = nil;
    self.player = nil;
}

// Listen for player authentication required
- (void)zenderPlayer:(ZenderPlayer *)zenderPlayer onZenderPlayerAuthenticationRequired:(NSDictionary *)payload {
    // do something to login the user
}

// Listen for Quiz Share codes
- (void)zenderPlayer:(ZenderPlayer *)zenderPlayer onZenderQuizShareCode:(NSDictionary *)payload {
    // TODO
}


@end
